usaspending_api/common/serializers.py

```diff
@@ -7,83 +7,127 @@ class LimitableSerializer(serializers.ModelSerializer):
-                allowed = set(include_fields)
-                existing = set(self.fields.keys())
-                for field_name in existing - allowed:
-                    try:
-                        # If we have a coded field, always include its description
-                        if field_name.split("_")[-1] == "description" and "_".join(field_name.split("_")[:-1]) not in allowed:
-                            continue
-                        self.fields.pop(field_name)
-                    except KeyError:
-                        # Because we're not currently handling nested serializer field
-                        # limiting pass-down, this can happen due to the context pass down
-                        pass
-
-            elif current_viewset and current_viewset.action == 'retrieve':
-                # The view has specifically asked that all fields should
-                # be returned (for example, when the request url follows
-                # the item/{pk} pattern)
-                return
-
-            else:
-                try:
-                    include_fields = self.Meta.model.get_default_fields(path=request._request.path_info)
-                    allowed = set(include_fields)
-                    existing = set(self.fields.keys())
-                    for field_name in existing - allowed:
-                        self.fields.pop(field_name)
-                except AttributeError:
-                    # We don't have get default fields available
-                    pass
+        # Now we alter our own field sets
+        # We must exclude before include to avoid conflicts from user error
+        for field_name in exclude_fields:
+            self.fields.pop(field_name)
+
+        if len(include_fields) > 0:
+            allowed = set(include_fields)
+            existing = set(self.fields.keys())
+            for field_name in existing - allowed:
+                # If we have a coded field, always include its description
+                if field_name.split("_")[-1] == "description" and "_".join(field_name.split("_")[:-1]) not in allowed:
+                    continue
+                self.fields.pop(field_name)
+
+    # Returns a list of child names for fields in the field list
+    # This is necessary so that if a user requests "recipient__recipient_name"
+    # we also include "recipient" so that it is serialized
+    def identify_missing_children(self, model, fields):
+        children = []
+        model_fields = [f.name for f in model._meta.get_fields()]
+        for field in fields:
+            split = field.split("__")
+            if len(split) > 0 and split[0] in model_fields and split[0] not in fields:
+                children.append(split[0])
+
+        return children
+
+    # Takes a child's name and a list of fields, and returns a set of fields
+    # that belong to that child
+    def get_child_fields(self, child, fields):
```

Comment:
I think its important to test the smaller bits of logic as long is it makes sense to test them, as well as testing them incorporated into the larger scope. I can write some unit tests for these (And fix up the doc strings) when I loop back around.